File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n, 1);

        // Calculate left products
        int left_product = 1;
        for (int i = 0; i < n; ++i) {
            ans[i] = left_product;
            left_product *= nums[i];
        }

        // Calculate right products and multiply with left products
        int right_product = 1;
        for (int i = n - 1; i >= 0; --i) {
            ans[i] *= right_product;
            right_product *= nums[i];
        }

        return ans;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-2.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()  # Sort the input array
        ans = []  # Store the result triplets
        
        for i in range(len(nums) - 2):  # Iterate through the array
            # Skip duplicates for the current element
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            j, k = i + 1, len(nums) - 1  # Two pointers
            
            while j < k:
                sum1 = nums[i] + nums[j] + nums[k]
                
                if sum1 == 0:
                    ans.append([nums[i], nums[j], nums[k]])  # Store the triplet
                    
                    # Move the pointers and skip duplicates
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
                
                elif sum1 < 0:
                    j += 1  # Increase the left pointer to get a larger sum
                else:
                    k -= 1  # Decrease the right pointer to get a smaller sum
        
        return ans
